<a href="https://colab.research.google.com/github/ssanchezgoe/diplomado_udea/blob/master/flask_colab_primera_app.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


<center>
    <img src="https://github.com/ssanchezgoe/flask_vscode_tutorial/blob/master/imagenes/flask_colab_logo.png?raw=true" alt="centered image" />
</center>

Google Colab nos proporciona una máquina virtual (Virtual Machine- VM), por lo que no tenemos acceso a un localhost, como si lo tenemos en un pc propio cuando ejecutamos un servidor web local. Lo que se puede hacer en este caso is acudir a un URL público usando la librería de python **flask-ngrok**. 



---
**ngrok:** es una herramiente que permite el acceso de cualquier persona en internet al contenido almacenado en un servidor local al generar un URL dinámicamente. Esta herramienta resulta útil en el caso que, por ejemplo, trabajemos con desarrolladores de forma remota y se pretende mostrar los 

Ngrok es una herramienta que permite acceder nuestro servidor local a cualquier persona en internet con la que compartamos una url generada dinamicamente , esto es muy útil por ejemplo cuando necesitamos mostrar avances constantemente en sitios que se encuentran en etapa de desarrollo o cuando trabajamos con un equipo de desarrolladores de forma remota .


NGROK nos permite realizar esto sin hacer ninguna configuración extra en el router o firewall simplemente basta con bajar la pequeña aplicación

---


Instalemos dicha librería primero:

In [4]:
!pip install flask-ngrok

# **Primer programa en `flask`**:

In [5]:
from flask_ngrok import run_with_ngrok
from flask import Flask

app = Flask(__name__)

run_with_ngrok(app)   #Ejecución de ngrok una vez la app está en marcha

# Rutas y vistas.
@app.route("/")
def home():
    return "<h1>Ejecución de Flask on Google Colab!</h1>"
  
if __name__ == '__main__':
 app.run()

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)


 * Running on http://58fd7a8e628a.ngrok.io
 * Traffic stats available on http://127.0.0.1:4040


127.0.0.1 - - [17/Aug/2020 22:53:42] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [17/Aug/2020 22:53:43] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [17/Aug/2020 22:53:47] "GET / HTTP/1.1" 200 -


Donde se obtiene, en el segundo URL, un URL público seguro al que puede acceder una persona desde internet.


## **Explicación del código:**

La explicación del código anterior se limitará a las sentencias en Flask y no al puente construido con **ngrok** para poder ejecutar flask en el colab. El código que puede ser ejecutado en un ordenador local con `flask` sería el siguiente:


```
from flask import Flask

app = Flask(__name__)

@app.route("/")
def home():
    return "<h1>Ejecución de Flask on Google Colab!</h1>"

if __name__ == '__main__':
 app.run()
```

###**Inicialización:**

Todas las applicaciones en `Flask` deben crear una instancia de aplicación. El servidor web envía todas las solicitudes que recibe de los clientes a esta instancia u objeto creado mediante un protocolo llamado *Interfaz de puerta de enlace del servidor web* (WSGI). La instancia es entonces un objeto de la clases Flask que se crea mediante la sentencia:

```
app = Flask(__name__)
```

El único argumento que se requiere al usar el constructor de la clase `Flask` es el nómbre del módulo o paquete principal; en la mayoría de aplicaciones la variable `__name__` representa el valor correcto. El argumento `__name__` determina, en este caso, la ruta raiz de la aplicación, de tal forma que, posteriormente, ésta pueda encontrar archivos de recursos relacionados con la ubicación de la aplicación. Sobre el argumento `__name__` se debe puntualizar lo siguiente:

- Cuando el intérprete lee un archivo de código, ejecuta todo el código global que se encuentra en él. Esto implica crear objetos para toda función o clase definida y variables globales.

- Todo módulo (archivo de código) en Python tiene un atributo especial llamado `__name__` que define el espacio de nombres en el que se está ejecutando. Es usado para identificar de forma única un módulo en el sistema de importaciones.

- Por otra parte, la cadena `'__main__'` es el nombre del ámbito en el que se ejecuta el programa principal.

- El intérprete pasa el valor del atributo `__name__` a la cadena `'__main__'` si el módulo se está ejecutando como programa principal (ejecutando el presente colab).

- Si el módulo no es llamado como programa principal, sino que es importado desde otro módulo, el atributo `__name__` tendrá como valor el nombre de dicho modulo. Al ejecutar en la terminal `Flask.__name__` se obtentra como resultado la cadena `Flask`.

###**Funciones de vista y rutas**

Un cliente (por ejemplo un navegador web), debe enviar una solicitud al servidor web, el cual la envía a la instacia de la aplicación `Flask`. La instancia debe saber qué código se debe ejecutar para cada URL solicitada, por lo que mantiene una asignación de las URL a las funciones de Python. La asociación entre una URL y la función que la manipula se le conoce con el nombre de *ruta*. Para definir una ruta en `flask` es a través del decorador `app.route` contenido en el objeto `app` y que regista la función decorada como ruta. La forma de invocar y registrar la función como ruta es la siguiente:



```
@app.route("/")
def home():
    return "<h1>Ejecución de Flask on Google Colab!</h1>"
```




---
**Decorador:** En `python`, un decorador es una característica estándar que puede modificar el comportamiento de una función de varias formas. Un patrón común es usar un decorador para registrar un función como un controlador para un evento.

---


En estas líneas de código, se registra la función `index()` como controlador de la aplicacion raíz. El valor que retorna esta función, que se conoce como **respuesta**, es lo que recibe el cliente. En el caso de un cliente representado por un navegador web, *la respuesta es el documento que se muestra al usuario*. Por otra parte, la función `index()` se denomina **función de vista**. En nuestro caso, la respuesta devuelta por esta función de vista es una cadena simple con contenido HTML, no obstante, puede adquirir formas más complejas. 

**Nota:** En este código las respuestas son embebidas en cadenas de caracteres que siguen un código HTML, lo cual es difícil de mantener. Más adelante se verá como generar respuestas a través de plantillas.

###**Inicio del servidor**

La instancia de la aplicación posee un método de ejecución que inicia el servidor web de desarrollo integrado:

    if __name__ == '__main__':
      app.run()


El booleano del condicional `__name__ == '__main__'` en `Python` se usa para asegurar que el servidor web de desarrollo solo se iniciará cuando se ejecute el script directamente. Cuando el script es importado por otro, se asume que el script principal iniciará un servidor diferente, por ende, la llamada `app.run()` será omitida.

**NOTA:** El servidor web  proporcionado por `Flask` no tiene como fin un uso de producción.

#**Reglas de variables**

Se puede adicionar seciones variales a un URL indicando las seciones como `<nombre_variable>`. En este caso, la función recibe `<nombre_variable>` como un argumento basado en una palabra clave. Opcionalmente, se puede usar un convertidor para especificar el tipo de argumento que se busca; para ello especificamos el tipo de la forma `<converter:variable_name>`. En la siguiente tabla, se definen los tipos de convertidores aceptados

Convertidor | Significado
--- |----
`string` | Acepta cualquier texto sin slash (definido por defecto).
`int`| Admite un entero positivo
`float` | Admite un valor en coma flotante positivo
`path` | Similar a un string, pero se adimiten slashes
`uuid` | Admite strings UUID

#**URLs únicos - Comportamiento en redirecionamiento**

Las siguientes dos formas de definir una ruta difieren según el uso del slash final



```
@app.route('/proyectos/')
def proyectos():
    return 'Página del proyecto'

@app.route('/about')
def about():
    return 'Acerca de la página'
```
The canonical URL for the projects endpoint has a trailing slash. It’s similar to a folder in a file system. If you access the URL without a trailing slash, Flask redirects you to the canonical URL with the trailing slash.

The canonical URL for the about endpoint does not have a trailing slash. It’s similar to the pathname of a file. Accessing the URL with a trailing slash produces a 404 “Not Found” error. This helps keep URLs unique for these resources, which helps search engines avoid indexing the same page twice.


# **Ciclo solicitud-respuesta**

A continuación se expondrá alguno de los aspectos de diseño de `Flask`.

##**Aplicaciones y contextos de solicitud:**

Cuando `Flask` recibe una solicitud de un cliente, se necesita crean algunos objetos que estarán disponibles para que la *función de vista* los manipule. Un ejemplo de dichos objetos es **objeto de solicitud**, que se enmascara la solicitud HTTP enviada por el cliente. 

La forma mas evidente en la cual `Flask` le puede dar acceso a una función de vista al objeto de solicitud es enviarlo como un argumento, pero esto requeriría que cada función de vista de la apliación tuviese un argumeto de más. Además, las cosas se complicarían si se considera que los objetos de solicitud no son los únicos objetos que las funciones de vista deben acceder para cumplir con una solicitud.

Para evitar saturar las funciones de vista con muchos argumentos, Flask usa el concepto de **contextos** para hacer que ciertos objetos sean accesibles globalmente temporalmente. Gracias a los contextos, se pueden escribir funciones de vista como la siguiente:



```
from flask import request

# Rutas
@app.route('/')

# Vista
def index():
  user_agent = request.headers.get('User-Agent')
  return '<p>Your browser is %s</p>' % user_agent
```

Nótese como en esta función de vista, la solicitud es usada como si fuese una variable global. En realidad, la solicitud no puede ser una variable global si consideramos que en un servidor multihilos, los hilos trabajan en solicitudes de diferentes clientes al mismo tiempo, de forma que a cada proceso se le debe "reservar" un objeto diferente en la solicitud. Los contextos le permiten a `Flask` hacer ciertas variables globales para que sean accesibles a un hilo sin interferir con otros.



---
**Hilo:** secuencia pequeña de instrucciones que se pueden administrar de forma independiente. En un proceso se pueden tener varios hilos activos al mismo tiempo, algunas veces, compartiendo recursos como memoria o identificadores de archivos. Los servidores web multihilos inician un subconjunto de hilos y seleccionan un hilo de ellos para manejar cada solicitud entrante.

---

En `Flask`, existen dos **contextos**: 

- Los contextos de aplicación.
- Los contextos de solicitud.

En la siguiente tabla se muestran las variables expuesta por cada uno de estos contextos.


Variable | Context | Descripción
--|--|--
`current_app` | Contexto de aplicacón | Instancia de aplicación para una aplicación activa
`g` | Contexto de aplicación | Objeto que la aplicación puede usar como almacenamiento temporal durante <br>la manipulación de una solicitud. Esta variable se reinicia en cada solicitud.
`request` | Contexto de solicitud | Objeto de solicitud, el cual enmascara los contenidos de una solicitud HTTP <br> enviada por el cliente.
`session` | Contexto de solicitud | Sesión de usuario que corresponde a un diccionario que la aplicación puede <br> usar para almacenar valores que deben tenerse en cuentra entre solicitudes.

Antes de enviar una solicitud, `Flask` activa la app y los contextos de solicitud, los cuales son eliminados cuando se tramíta la solicitud. Cuando opera el contexto de aplicación, las variables `current_app` y `g` están disponibles para el hilo; así mismo, cuado el contexto de solicitud opera, la solicitud y la sesión también están disponibles. Si se accede a cualquiera de estas variables sin tener alguna aplicación activa o contexto de solicitud, se fenera un error. Por ejemplo, veamos el error que se obtendrá sino se ha definido el contexto de aplicación si corremos, junto a l archivo de [este link](https://github.com/ssanchezgoe/flask_vscode_tutorial/blob/master/hello.py), las siguientes líneas de código :

In [5]:
!wget https://raw.githubusercontent.com/ssanchezgoe/flask_vscode_tutorial/master/hello.py

from hello import app
from flask import current_app

current_app.name

--2020-08-17 18:51:35--  https://raw.githubusercontent.com/ssanchezgoe/flask_vscode_tutorial/master/hello.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 247 [text/plain]
Saving to: ‘hello.py.1’

hello.py.1          100%[===================>]     247  --.-KB/s    in 0s      

2020-08-17 18:51:35 (13.1 MB/s) - ‘hello.py.1’ saved [247/247]



RuntimeError: ignored

Para evitar este error, debemos antes activar el contexto de la aplicación mediante la siguiente forma

In [7]:
from hello import app
from flask import current_app

app_ctx = app.app_context()
app_ctx.push()
print('Ahora si que obtenemos el nombre de la app en ejecución: ', current_app.name)
app_ctx.pop()

Ahora si que obtenemos el nombre de la app en ejecución:  hello


##**Solicitud de envío**

Cuando una aplicación recibe una solicitud de un cliente, necesita saber que función de vista llama para tramitarla. Para realizar esta tarea, `Flask` busca la URL proporcionada en la solucitud dentro del mapa URL de la aplicación. Este mapa es creado usando los decoradores de `app.route` o la versión equivalente (que no es un decorador) `app.add_url_rule()`.

Para ver el mapa URL en una aplicación `Flask`, podemos hacerlo al inspecionar el mapa creado para la aplicación `hello.py` que descargamos anterioremente. 

In [6]:
from hello import app
app.url_map

Map([<Rule '/' (GET, OPTIONS, HEAD) -> index>,
 <Rule '/static/<filename>' (GET, OPTIONS, HEAD) -> static>,
 <Rule '/user/<name>' (GET, OPTIONS, HEAD) -> user>])

Del anterior mapa se puntualiza lo siguiente:


- Las siguientes rutas son definidas por los decoradores de `app.route`:

  - / 
  - /user/<name> 

- `Flask` adiciona la ruta `/static/<filename>`para darle acceso a archivos estáticos (de los cuales se hablará más adelante). 

- Los elementos:

  - HEAD
  - OPTIONS
  - GET

  repreentan los **métodos** de solicitud que son empleados por la ruta. `Flask` adjunta a cada ruta  métodos de tal forma que diferentes métodos de solicitud enviados al mismo URL puedan ser empleados por diferentes funciones de vista. Los métodos `HEAD` y `OPTIONS` son empleados automáticamente por `Flask`, de tal forma que en la práctica se puede decir en esta aplicación las tres rutas en el mapa URL se adjuntan al método `GET`. Más adelande se estudiarán diferentes métodos de solicitud para rutas.

##**Request Hooks**

En ocaciones resulta útil ejecutar un código antes o después de que cada solicitud sea procesada. Por ejemplo, en el inicio de una solicitud, puede ser necesario crear una conexión con una base de datos, o autenticar al usuario que hace la solicitud. En lugar de duplicar el código que hace esta tarea en cada función de vista, `Flask` proporciona la opción de registrar funciones comunes las cuales se llaman antes o después de que se envíe una solicitud a una función de vista. 

Los request hooks son implementados como decoradores y existen cuatro decoradores soportados por `Flask:`

- `before_first_request`: Registro de una función para ejecutarse antes de que la primera solicitud sea tratada.
-  `before_request`: Registo de una función para ejecutarse antes de cada solicitud.
- `after_request`: Registro de una función para ejecutarse después de caada solicitud, siempre que no se hayan producido excepciones no controladas.
- `teardown_request`: Registro de una función para ejecutarse despues de cada solicitd, inclusi si se han producido excepciones no controladas. 

Un patrón común para compartir datos entre las funciones de enlace de solicitud y las funciones de visualización es usar el contexto global `g`. 

---
**Ejemplo:** un controlador `before_request` puede cargar el usuario que inició sesión desde la base de datos y almacenarlo en `g.user`. Posteriormente, cuando se invoca la función de visualización, puede acceder al usuario desde allí.

---


A lo largo del curso se mostrarán ejemplos de enlaces de solicitud.

##**Respuestas**

Cuando `Flask` invoca una función de vista, espera que su valor de retorno sea la respuesta a la solicitud. En la mayoria de casos corresponde a una cadena que es enviada al cliente como una página HTML. No obstante, el protocolo HTTP precisa de más que una cadena como respuesta a una solicitud. Una parte primordial de la respuesta HTTP es el **estatus del código**, el cual `Flask` fija por defecto en 200, el cual indica que la solicitud fue entregada satisfactoriamente.



Cuando una función de vista necesita responder con un código de estado diferente, puede agregar el código numérico como un segundo valor de retorno después del texto de respuesta. Por ejemplo, la siguiente función de vista devuelve un código de estado 400, el código de un error de solicitud incorrecta:

#**Prácticas**:

A continuación se desarrollará unas prácticas constituidas por problemas relacionados con la temática vista en este notebook. 

En el presente [link](https://flask.palletsprojects.com/en/1.1.x/) se brinda un material de apoyo que usaremos para repazar los conceptos vistos.

## **Pregunta 1:** 

El protocolo usado por `Flask` para pasar todas las solicitudes de los clientes a un objeto que los manipule es:

In [4]:
#@title
Protocolo = "WSGI" #@param ["WSGI", "HTTP", "WWW","N/A"]

if Protocolo == "WSGI":
  print('''La respuesta es correcta:
El servidor de puerta de enlace de la interfaz web (WSGI) 
es una simple convención de llamada para los servidores web 
para reenviar solicitudes a las aplicaciones web escritas en 
el lenguaje de programación Python .
''')
else:
  print('''Respuesta incorrecta:
Debes repazar el primer programa implementado en este notebook.
  ''')

La respuesta es correcta:
El servidor de puerta de enlace de la interfaz web (WSGI) 
es una simple convención de llamada para los servidores web 
para reenviar solicitudes a las aplicaciones web escritas en 
el lenguaje de programación Python .



##**Preguntas 2 a 5:** 

Para las preguntas 2 a 5 analice las siguientes líneas de código



```
from flask import Flask
app = Flask(arg)

@app.route('/')
def hola_mundo():
    return 'Hola, Mundo!'
```

In [ ]:
#@title **Pregunta 2. El argumento denotado por `arg` para crear una instancia de la clase `Flask` debe ser:**
arg = "None" #@param ["None", "__name__", "__main__","run"]

if arg == "__name__":
  print('''Respuesta correcta!

El primer argumento es el nombre del módulo o 
paquete de la aplicación. Si está usando un solo módulo,
debe usar el nombre porque, dependiendo de si se inició 
como aplicación o se importómcomo módulo, el nombre será 
diferente (__main__ frente al nombre de importación real). 
Esto es necesario para que Flask sepa dónde buscar 
plantillas, archivos estáticos, etc.
''')
else:
  print('''Respuesta incorrecta:
Debes repazar el primer programa implementado en este notebook.
  ''')



In [ ]:
#@title **Pregunta 3 (verdador o falso). La aplicación Flask definida, se divide, principalmente, en funciones de ruta y vista**
arg = "Verdadero" #@param ["Falso", "Verdadero"]

if arg == "Verdadero":
  print('''Respuesta correcta!

Las partes centrales en una aplicación web estan constituidas
por las rutas creadas a partir de las solicitudes realizadas 
por un usuario (como un navegador web) a funciones de vista 
(escritas en Python, en nuestro caso), y las funciones de vista 
en sí, encargadas de dar respuesta en forma de una visualización 
en el navegador, por ejemplo. 
''')
else:
  print('''Respuesta incorrecta:
Debes repazar el primer programa implementado en este notebook.
  ''')



In [ ]:
#@title **Pregunta 4. la sentencia `@app.route('/')` se encarga de:**
arg = "Asociar un URL a una funcion de vista de Python" #@param ["Nada", "Iniciar la visualizacion de la aplicacion", "Asociar un URL a una funcion de vista de Python","Definir un decorador vacio"]

if arg == "Asociar un URL a una funcion de vista de Python":
  print('''Respuesta correcta!

@app.route('/') define corresponde a un decorador en Python encar-
gado de asociar la un URL a la función de vista hola_mundo(). En 
este código se registra la función index() como el controlador del 
URL raiz de la aplicación.
''')
else:
  print('''Respuesta incorrecta:
Debes repazar el primer programa implementado en este notebook.
  ''')



In [ ]:
#@title **Pregunta 5. La función hola_mundo() es una función de vista encargada de:**
hola_mundo = "Definir una funcion de ruta" #@param ["Definir una funcion de ruta", "Generar una respuesta que corresponde a un string", "Inicializar la aplicacion","Finalizar la ejecucion del programa"]

if hola_mundo == "Generar una respuesta que corresponde a un string":
  print('''Respuesta correcta!

La función hola_mundo() se encarga de generar una respuesta
a la petición de un cliente para el URL raiz de la aplica-
ción.

Nota: Las respuestas de strings embebidas en python no son la
herramientas óptimas para generar código HTML. Más adelante 
veremos como usar plantillas para generar respuestas.
''')
else:
  print('''Respuesta incorrecta:
Debes repazar el primer programa implementado en este notebook.
  ''')



##**Problema 6.**

Para la aplicación del principio del notebook



```
from flask_ngrok import run_with_ngrok
from flask import Flask

app = Flask(__name__)

run_with_ngrok(app)   #Ejecución de ngrok una vez la app está en marcha

# Rutas y vistas.
@app.route("/")
def home():
    return "<h1>Ejecución de Flask on Google Colab!</h1>"
  
if __name__ == '__main__':
 app.run()
```



defina una ruta dinámica de la forma `'/user/<name>'` para personalizar el URL. Defina además una función de vista `user(name)`, en donde se le de la bienvenida al usuario almacenado en la variable `name`

In [ ]:
#@title **Solución al problema**
from flask_ngrok import run_with_ngrok
from flask import Flask
app = Flask(__name__)
 
run_with_ngrok(app)

@app.route("/")
def home():
    return "<h1>Ejecución de Flask on Google Colab!</h1>"

@app.route('/user/<name>')
def user(name):
  return "<h1>Bienvenido %s a esta clase!</h1>" % name

print('''Ayuda:

Para ejecutar esta aplicación, debe copiar el URL
suministrado por ngrok y adicionar la ruta con el usuario,
es decir, URL/user/nombre_usuario.

''')

if __name__ == '__main__':
 app.run()